flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/file/lsm/BinaryMemTable.java

```diff
@@ -0,0 +1,222 @@
+import org.apache.flink.table.planner.codegen.ProjectionCodeGenerator;
+import org.apache.flink.table.planner.codegen.sort.SortCodeGenerator;
+import org.apache.flink.table.planner.plan.utils.SortUtil;
+import org.apache.flink.table.runtime.generated.GeneratedNormalizedKeyComputer;
+import org.apache.flink.table.runtime.generated.GeneratedProjection;
+import org.apache.flink.table.runtime.generated.GeneratedRecordComparator;
+import org.apache.flink.table.runtime.generated.NormalizedKeyComputer;
+import org.apache.flink.table.runtime.generated.Projection;
+import org.apache.flink.table.runtime.generated.RecordComparator;
+import org.apache.flink.table.runtime.operators.aggregate.BytesHashMapSpillMemorySegmentPool;
+import org.apache.flink.table.runtime.operators.sort.BinaryKVInMemorySortBuffer;
+import org.apache.flink.table.runtime.typeutils.BinaryRowDataSerializer;
+import org.apache.flink.table.runtime.typeutils.RowDataSerializer;
+import org.apache.flink.table.runtime.util.collections.binary.BytesHashMap;
+import org.apache.flink.table.runtime.util.collections.binary.BytesMap;
+import org.apache.flink.table.storage.file.utils.DualIterator;
+import org.apache.flink.table.types.logical.BigIntType;
+import org.apache.flink.table.types.logical.LogicalType;
+import org.apache.flink.table.types.logical.RowType;
+import org.apache.flink.table.types.logical.TinyIntType;
+import org.apache.flink.util.MutableObjectIterator;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+public class BinaryMemTable implements MemTable {
+
+    private final BytesHashMap table;
+
+    // keyType + long + byte
+    private final RowType storeKeyType;
+    private final RowType valueType;
+    private final int userKeyLength;
+    private final RowDataSerializer storeKeySerializer;
+    private final RowDataSerializer valueSerializer;
+
+    private final GenericRowData reuseMetaRow;
+    private final JoinedRowData reuseStoreRow;
+
+    // for sort binary buffer
+    private final NormalizedKeyComputer normalizedKeyComputer;
+    private final RecordComparator keyComparator;
+    private final Projection<RowData, BinaryRowData> projection;
+
+    public BinaryMemTable(RowType userKeyType, RowType valueType, long maxMemSize) {
+        List<LogicalType> types =
+                userKeyType.getFields().stream()
+                        .map(RowType.RowField::getType)
+                        .collect(Collectors.toList());
+        types.add(new BigIntType());
+        types.add(new TinyIntType());
+        LogicalType[] keyTypes = types.toArray(new LogicalType[0]);
+        LogicalType[] valTypes =
+                valueType.getFields().stream()
+                        .map(RowType.RowField::getType)
+                        .toArray(LogicalType[]::new);
+        this.table = new BytesHashMap(null, null, maxMemSize, keyTypes, valTypes);
```

Comment:
Intead of using `BytesHashMap`, maybe we can concat them as a Row just like `SstFileWriter`.